### PR TITLE
THE TEST RELEASE

### PR DESCRIPTION
**Pull request by ian1dunn**
# Overview
### Added
- A timer displaying the elapsed time of the test to the top of the control window.

### Added
- All test image assets
- Abort test button
- Audio cues for blinking and transition tests
- Several new test timing variables to config

### Changed
- Changed from each test having its own class to a template-based test system:
  - All tests extend a base `TestThread` class
  - `BlinkTest`, `ConstantTest`, and `TransitionTest` are three different test types with three different behaviors
- Timestamps are now a constant millisecond value from an arbitrary point.
  - See [the LSL documentation](https://labstreaminglayer.readthedocs.io/info/faqs.html?highlight=timestamp#lsl-local-clock) for the reasoning behind this.

### Fixed
- All test images now appear with transparent backgrounds
- Labeling in saved CSV files

### Removed
- `postprocessing` module as we no longer display graphs after collection

# Testing
To test this PR, please copy and paste this into your review:
- [ ] Blink
  - [ ] Abort button
- Constant Tests
  - [ ] Random test 1
  - [ ] Random test 2
  - [ ] Abort button
- Transition Tests
  - [ ] Random test 1
  - [ ] Random test 2
  - [ ] Abort button

Look at the recorded data in the CSV. Do the labels seem to match up to the test states? Is all of the data being logged? Are the millisecond timestamps accurate?

# Notes
- Test times (with default config values):
  - Blink: 145sec ~= 2.4min
  - Constant: 260sec ~= 4.3min
  - Transition: 305sec ~= 5.08min
  - Estimated time for one trial per test: `145sec * 1 + 260sec * 10 + 305sec * 7 = 4880sec = 81.3min = 1hr 20min`

### CODE REVIEW

File: LSL.py [hunk 1]
```diff
@@ -68,16 +81,18 @@ def start_label(event: str):
         """
         Function to start labelling each data frame until stop_label() is called
         """
-        LSL.collection_label = event
-        print(f"Labeling Data: {event}")
+        if not event == LSL.collection_label:
```
Comment (Matt-London) [inline LSL.py:84]: Maybe `if event != LSL.collection_label:`

File: LSL.py [hunk 2]
```diff
@@ -144,7 +157,8 @@ def __save_collected_data(path: str):
 
                 # Convert collected data to a DataFrame, format with columns above, and write to CSV
                 df = pd.DataFrame(LSL.collected_data[stream_type], columns=columns)
-                df['Timestamp'] = pd.to_datetime(df['Timestamp'], format='mixed')#'%Y-%m-%d %H:%M:%S.%f')
+                # Formatting breaks everything for some reason:
+                # df['Timestamp'] = pd.to_datetime(df['Timestamp'], format='%Y-%m-%d %H:%M:%S.%f')
```
Comment (Matt-London) [inline LSL.py:161]: If this is old code it should probably be deleted. Commented out code is generally avoided.

File: main.py [hunk 3]
```diff
@@ -9,14 +13,30 @@ class DataCollectorApp:
...
-        test.start()
+        # Dynamically import the test from tests package & construct it w/ no parameters
+        class_name = f"{test_type}Test"
+        test_class = getattr(importlib.import_module(f"tests.{class_name}"), class_name)
+
+        if test_type == "Transition":
+            assets = config.TESTS[test_type][test_name]
+            test = test_class(test_name, os.path.join('.', 'assets', assets[0]), os.path.join('.', 'assets', assets[1]))
+        elif test_type == "Constant":
+            # asset = config.TESTS[test_type][test_name]  # No picture needed as of now
```
Comment (Matt-London) [inline main.py:30]: Unless this will be used later it should probably be removed

File: tests/TestGUI.py [hunk 4]
```diff
@@ -83,122 +109,124 @@ def confirm_current_test(test_data_path: str) -> bool:
...
+
+        # Start the timer
+        TestGUI.timer_thread = threading.Thread(target=TestGUI.update_timer)
+        TestGUI.timer_thread.start()
+
+        TestGUI.test_start_time = time.time()  # Record the start time
 
     #
     # HELPER METHODS
     #
```
Comment (Matt-London) [inline tests/TestGUI.py:147]: This is opinion based but I personally don't like labeling file sections like this, especially if it isn't used across all files

File: tests/TestGUI.py [hunk 5]
```diff
@@ -83,122 +109,124 @@ def confirm_current_test(test_data_path: str) -> bool:
...
 
     @staticmethod
-    def __show_data_and_confirm(test_data_path: str):
+    def update_timer():
```
Comment (Matt-London) [inline tests/TestGUI.py:150]: If it is a "helper function" only used within the file/class and should be private it should start with two underscores like this:
```
def __update_timer():
```

File: tests/TestGUI.py [hunk 6]
```diff
@@ -83,122 +109,124 @@ def confirm_current_test(test_data_path: str) -> bool:
...
-        popup = tk.Toplevel()
-        popup.wm_title("Data Confirmation")
-        TestGUI.disable_close_button(popup)
+        while True:
+            # Check for messages in the queue
+            try:
+                message = TestGUI.timer_queue.get(block=False)
+                if message == "stop":
+                    # Stop the timer
+                    break
```
Comment (Matt-London) [inline tests/TestGUI.py:160]: Instead message could be initialized above the while loop and the condition could be `while message != "stop"`. It accomplishes the same thing but makes it a bit more clear as to when the loop quits

File: tests/TestGUI.py [hunk 7]
```diff
@@ -266,3 +295,42 @@ def __set_window_geometry(window, left_side=True):
...
+        Helper function to place an image in the middle of the display window. Returns the ID of the image object
+        for destruction.
+
+        :param image: The Tkinter image to place
+        """
+        if TestGUI.current_display_element is not None:
+            TestGUI.destroy_current_element()
+
+        x = TestGUI.display_canvas.winfo_width() / 2
+        y = TestGUI.display_canvas.winfo_height() / 2
```
Comment (Matt-London) [inline tests/TestGUI.py:311]: Should this be `//` to make sure it's an int?

File: tests/TestGUI.py [hunk 8]
```diff
@@ -266,3 +295,42 @@ def __set_window_geometry(window, left_side=True):
...
+        Helper function to place text in the middle of the display window. Returns the ID of the text object
+        for destruction.
+
+        :param text: The text to place
+        """
+        if TestGUI.current_display_element is not None:
+            TestGUI.destroy_current_element()
+
+        x = TestGUI.display_canvas.winfo_width() / 2
+        y = TestGUI.display_canvas.winfo_height() / 2
```
Comment (Matt-London) [inline tests/TestGUI.py:326]: Same thing here potentially?

File: tests/TestThread.py [hunk 9]
```diff
@@ -13,44 +14,56 @@ class TestThread(threading.Thread):
...
-        TestGUI.display_window.after(config.TEST_DURATION, self.stop)
+        self.run_test()
 
         return  # End thread
 
+    def run_test(self):
+        """
+        Main loop that runs and schedules the next iteration of the test
+        """
+        pass  # Override me to implement a new test 💀
```
Comment (Matt-London) [inline tests/TestThread.py:65]: Not the emoji 💀

**Review by Matt-London — APPROVED** [review]
Looks good overall. There's like three small things I left comments about but nothing is necessary.

File: LSL.py [hunk 10]
```diff
@@ -112,16 +127,15 @@ def __collect_data():
         for real-time data processing. Uses StreamInlet.time_correction() to convert LSL to system timestamps using a
         constantly updated offset. The precision of these estimates should be below 1 ms (empirically within +/-0.2 ms).
         """
-
         while LSL.collecting:
             for stream_type, stream in LSL.streams.items():
-                data_row = {'Timestamp': None, 'Label': "" if not LSL.collection_label else LSL.collection_label}
+                data_row = {'Timestamp': None, 'Label': "Resting" if not LSL.collection_label else LSL.collection_label}
```
Comment (Alexandr5656) [inline LSL.py:132]: I want to change it to no event instead of resting

File: tests/TestGUI.py (hunk 5 above)
Comment (Matt-London) [reply tests/TestGUI.py:150]: I meant single underscores

File: tests/TestGUI.py [hunk 11]
```diff
@@ -64,17 +90,17 @@ def add_test(test_name: str, test_lambda):
...
-    def confirm_current_test(test_data_path: str) -> bool:
+    def confirm_current_test() -> bool:
         """
         Run the test and prompt the user to confirm or deny the data.
-
-        :param test_data_path: The path to the current test data FOLDER.
         """
         # Confirm data
-        TestGUI.__show_data_and_confirm(test_data_path)
+        TestGUI.__show_data_and_confirm()
```
Comment (Alexandr5656) [inline tests/TestGUI.py:98]: We dont show the data anymore is this just to bring a pop up?

**Review by Alexandr5656 — COMMENTED** [review]
I chew gum and eat ass and im all out of ass

File: LSL.py (hunk 10 above)
Comment (ian1dunn) [reply LSL.py]: Added config variable

File: tests/TestThread.py (hunk 9 above)
Comment (ian1dunn) [reply tests/TestThread.py:65]: 💀

File: tests/TestGUI.py (hunk 6 above)
Comment (ian1dunn) [reply tests/TestGUI.py:199]: Would this be thread-safe? I added a TODO, @mml2034 this could possibly be something to look into for a future release.

**Pull request merged**